### PR TITLE
Stability and improvement 

### PR DESCRIPTION
**Pull request by jbsky**
The fixes are mainly for the WRT1900X and WRT1200AC.

It contains:

* AMPDU Optimization of priority calculation. (8864/8897)
* 8864/8897 fix wpa3
* Add cypher suite to driver capabilities (all chip)
* Alignment of pcie_tx_ctrl* (all chip)
* "feature" normalization for all chips (all chip)
* Add modprobe parameters (all chip)
* fix monitoring (all chip)
* Code separation by chipset (8864/8897 <=> 8997)
* dump_prob decommissioning (8864/8897/8997)
* fix amsdu high ping latency (8864/8897/8997)
* drop debug info in hostcmd_get_hw_spec() (8864/8897)
* Add wcb_base in debug info (8864) 
* Rewrite AMSDU packets (8864/8897/8997)
* debug rewrite output mwl_debugfs_sta_read  (all chip)
* Improved encryption interoperability (8864/8897/8997)
* factorization encrypted packet test (8864/8897/8997)
* Change 88W8864 firmware to 7.2.9.27 (8864/8897)
* Fix the AMPDU session lifecycle (8864/8897/8997)
* Remove the tx done packets mechanism (8864/8897)

Please check and take into account these expected changes.

Thank you in advance,

with kind regards,

Julien Blais

### CODE REVIEW

**Comment by jbsky** [comment]
Hello

It's pretty nice now and it's stable and robust with this version with WPA 3 ;).

```
root@wrt1900acv1:~# cat /sys/kernel/debug/ieee80211/phy{0,1}/mwlwifi/{ampdu,sta}
debug_ampdu: 0
=================
stream|idx|state|       macaddress|tid|pkts|remain|start time
       macaddress|aid|ampdu|amsdu|  wds|ba_hist|amsdu_cap|  ht cap, ampdu,         rx_mask|   vht_cap|           mcs|rx bw,nss|tdls|init|wme|mfp
70:03:9f:3d:00:d0|  1| true|false|false|disable|     0x00|0x00102c, 00-07,0xff, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  0
70:03:9f:3d:2e:06|  3| true|false|false|disable|     0x00|0x00102c, 00-07,0xff, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  0
70:03:9f:3d:2e:84|  4| true|false|false|disable|     0x00|0x00102c, 00-07,0xff, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  0
a8:80:55:b7:31:08|  8| true|false|false|disable|     0x00|0x00002c, 00-07,0xff, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  1
fc:67:1f:10:d5:60|  9|false|false|false|disable|     0x00|0x000000, 00-00,0x00, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  0
38:1f:8d:46:31:8a|  5| true|false|false|disable|     0x00|0x00002c, 00-07,0xff, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  0
38:1f:8d:45:74:19|  6|false|false|false|disable|     0x00|0x000000, 00-00,0x00, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  0
38:1f:8d:19:e7:cb|  2|false|false|false|disable|     0x00|0x000000, 00-00,0x00, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  0
ec:0b:ae:84:c5:bd|  7|false|false|false|disable|     0x00|0x000000, 00-00,0x00, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  1
38:1f:8d:46:b6:ab| 10|false|false|false|disable|     0x00|0x000000, 00-00,0x00, 00, 00, 00|0x00000000|0x0000, 0x0000|rx  0,  1|   0|   0|  1|  0
debug_ampdu: 0
=================
stream|idx|state|       macaddress|tid|pkts|remain|start time
     0|  0|    3|00:45:e2:31:91:d1|  0|  47|   199|659
     1|  1|    3|44:85:00:23:73:f0|  0|  13|   596|467
       macaddress|aid|ampdu|amsdu|  wds|ba_hist|amsdu_cap|  ht cap, ampdu,         rx_mask|   vht_cap|           mcs|rx bw,nss|tdls|init|wme|mfp
14:c1:4e:3b:f5:10|  1| true|false|false|disable|     0x00|0x00006f, 03-00,0xff, 00, 00, 00|0x33c07030|0xfffe, 0xfffe|rx  2,  1|   0|   0|  1|  0
00:45:e2:31:91:d1|  2| true| true|false|disable|     0x01|0x00006e, 02-00,0xff, 00, 00, 00|0x33c07020|0xfffe, 0xfffe|rx  2,  1|   0|   0|  1|  1
44:85:00:23:73:f0|  3| true| true|false|disable|     0x01|0x000067, 03-05,0xff, ff, 00, 00|0x038178b0|0xfffa, 0xfffa|rx  2,  2|   0|   0|  1|  1
a4:45:19:3e:8e:45|  4| true|false|false|disable|     0x00|0x00006f, 03-00,0xff, 00, 00, 00|0x03807030|0xfffe, 0xfffe|rx  2,  1|   0|   0|  1|  1
```
@kaloz 
This version was closely tested with 2 devices, wrt1900ac and wrt3200ac, and no problems were reported.

# FEATURES
While searching the git repository, I found the following features. Can you provide a detailed description of each of them?
```
#define HW_SET_PARMS_FEATURES_MAX_AMSDU_SIZE_4K       0x00000001
#define HW_SET_PARMS_FEATURES_MAX_AMSDU_SIZE_8K       0x00000002
#define HW_SET_PARMS_FEATURES_IMPLICITE_AMPDU_BA      0x00000004
#define HW_SET_PARMS_FEATURES_DISABLE_MBSS            0x00000008
#define HW_SET_PARMS_FEATURES_HOST_BEACON             0x00000010
#define HW_SET_PARMS_FEATURES_HOST_PROBE_RESP         0x00000020
#define HW_SET_PARMS_FEATURES_HOST_POWER_SAVE         0x00000040
#define HW_SET_PARMS_FEATURES_HOST_ENCR_DECR_MGMT     0x00000080
#define HW_SET_PARMS_FEATURES_HOST_INTRA_BSS_OFFLOAD  0x00000100
#define HW_SET_PARMS_FEATURES_HOST_INTRA_IV_OFFLOAD   0x00000200
#define HW_SET_PARMS_FEATURES_HOST_ENCR_DECR_FRAME    0x00000400
```
# MESH
I've also been trying to get MESH to work and I've got good news and bad news.
The good is that MESH works, the bad is only without encryption.

With a ping request from a capable and functional MESH card to the 8864 chip. The chip decodes correctly. We see an ARP request, and it stops there, since our emitter doesn't know the arp address, it can't ping.
image

It soon becomes clear that the problem lies with the encryption on the 8864 side.
One surprising thing, which I didn't investigate, is that the seqnum doesn't match.

The QOS sets MESH to 1, but there's no mesh header.
In fact, the mesh header part is passed to the encryption side in data[0].

image

A quick reminder about the DMA header
```
/* DMA header used by firmware and hardware. */
struct pcie_dma_data {
	__le16 fwlen;
	struct ieee80211_hdr wh;
	char data[0];
} __packed;
```
The excel representation below shows that the QOS part is passed separately and that a DMA header is set up so that it can be translated and sent correctly by the firmware.
In other words, a header with 4 addresses.

image

Perhaps I'm mistaken or lacking in precision, hardware encryption is always from data[0] + IV Header.

Is an encryption offset feature planned?
If Santa is listening.

To conclude a little about the year of development I've been able to do. Today, it's pretty stable and robust with this version.

We realize that MU MIMO is sorely lacking.
=> if a device performs a flow test with iperf (1x1), the chip is unable to establish another flow in 1x1 or 2x2.

Best regards,

Julien Blais

**Comment by kaloz** [comment]
Hi Julien,

Merged, thank you. I've also sent you a mail regarding the questions you had :)

**Comment by johnnyxwan** [comment]
Just saw this amazing merge and it basically revived my good old WRT1900. Appreciated your work and thanks! I have WRT3200 and WRT32X on my hand as well, it would be fantastic if they can be fixed as well. Let me know what I can do to assist if needed.

**Comment by jbsky** [comment]
Thanks, that's nice, but there's nothing practical with a remote user.
To test effectively, I'd need a WRT32.
Yes, I'd like to finish the job, but I don't want to put any money on my side. The material still costs a certain amount.
It's not for today. Sorry about that.

**Comment by kaloz** [comment]
> Thanks, that's nice, but there's nothing practical with a remote user.
> 
> To test effectively, I'd need a WRT32.
> 
> Yes, I'd like to finish the job, but I don't want to put any money on my side. The material still costs a certain amount.
> 
> It's not for today. Sorry about that.

@jbsky : I've sent you an email about this, but got no answer - maybe it ended up in spam or something. I would like to ask you to send me one and I'll help you out with some hw.

**Comment by jbsky** [comment]
> > Thanks, that's nice, but there's nothing practical with a remote user.
> > To test effectively, I'd need a WRT32.
> > Yes, I'd like to finish the job, but I don't want to put any money on my side. The material still costs a certain amount.
> > It's not for today. Sorry about that.
> 
> @jbsky : I've sent you an email about this, but got no answer - maybe it ended up in spam or something. I would like to ask you to send me one and I'll help you out with some hw.

Hello @kaloz ,

I've e-mailed you back.

Best regards,

**Comment by johnnyxwan** [comment]
> Thanks, that's nice, but there's nothing practical with a remote user. To test effectively, I'd need a WRT32. Yes, I'd like to finish the job, but I don't want to put any money on my side. The material still costs a certain amount. It's not for today. Sorry about that.

Totally understand and it is rather unreasonable for you to take your own money doing this stuff. What if we can figure out the amount it takes (hardware cost + shipping fee + perhaps some coffee), and then we try to raise fund for it? I am pretty sure a lot of people would happily support it rather than paying extra to buy a less capable router out there.

**Comment by nmschulte** [comment]
I as well have a WRT3200ACM I'd be willing to lend to the cause.  Thank you for your support about this issue!
